cookbooks/anisotropic_viscosity/av_material.cc

```diff
@@ -0,0 +1,858 @@
+            assemblers.stokes_preconditioner[i] = std_cxx14::make_unique<Assemblers::StokesPreconditionerAnisotropicViscosity<dim> > ();
+			}
+
+		for (unsigned int i=0; i<assemblers.stokes_system.size(); ++i)
+			{
+			if (dynamic_cast<Assemblers::StokesIncompressibleTerms<dim> *>(assemblers.stokes_system[i].get()) != nullptr)
+            assemblers.stokes_system[i] = std_cxx14::make_unique<Assemblers::StokesIncompressibleTermsAnisotropicViscosity<dim> > ();
+			}
+		}
+
+	template <int dim>
+	void
+	AV<dim>::
+	initialize()
+		{
+		this->get_signals().set_assemblers.connect (std::bind(&AV<dim>::set_assemblers,
+                                                          std::cref(*this),
+                                                          std::placeholders::_1,
+                                                          std::placeholders::_2));
+		AssertThrow((dim==2),
+		                ExcMessage("For now this works only in 2D"));
+
+		}
+	template <int dim>
+	void
+	AV<dim>::evaluate (const MaterialModel::MaterialModelInputs<dim> &in,
+                      MaterialModel::MaterialModelOutputs<dim> &out) const
+	{
+
+      MaterialModel::AnisotropicViscosity<dim> *anisotropic_viscosity =
+      out.template get_additional_output<MaterialModel::AnisotropicViscosity<dim> >();
+
+      Simple<dim>::evaluate(in,out);
+
+    AssertThrow((this->introspection().compositional_name_exists("gamma")),
+                ExcMessage("AV material model only works if there is a compositional field called gamma."));
+    AssertThrow(this->introspection().compositional_name_exists("ni"),
+                ExcMessage("AV material model only works if there is a compositional field called ni."));
+    AssertThrow(this->introspection().compositional_name_exists("nj"),
+                ExcMessage("AV material model only works if there is a compositional field called nj."));
+    if (dim == 3)
+      AssertThrow(this->introspection().compositional_name_exists("nk"),
+                  ExcMessage("AV material model only works if there is a compositional field called nk."));
+
+    const unsigned int c_idx_gamma = this->introspection().compositional_index_for_name("gamma");
+
+    std::vector<unsigned int> c_idx_n;
+    c_idx_n.push_back (this->introspection().compositional_index_for_name("ni"));
+    c_idx_n.push_back (this->introspection().compositional_index_for_name("nj"));
+    if (dim == 3)
+      c_idx_n.push_back (this->introspection().compositional_index_for_name("nk"));
+
+    // Get the grad_u tensor, at the center of this cell, if possible.
+
+    std::vector<Tensor<2,dim> > velocity_gradients (in.position.size());
+    if (in.current_cell.state() == IteratorState::valid)
+      {
+    	std::vector<Point<dim> > quadrature_positions(in.position.size());
+    	          for (unsigned int i=0; i < in.position.size(); ++i)
+    	            quadrature_positions[i] = this->get_mapping().transform_real_to_unit_cell(in.current_cell, in.position[i]);

```

Comment:
This procedure is actually quite inefficient. There are easier ways to compute all of this. I suspect that you haven't done this yourself but copied it from somewhere. If you let me know where that is, then I can fix it there and point you at it.

Reply:
It was copied from visco_plastic.cc (line 460). Based on a quick search, it is also in the viscoelastic.cc